Backend exercise

mpliego opened the pull request.

mpliego-travelperk committed:
Added Django base structure & Dockerfile.

mpliego-travelperk committed:
New docker-compose and Makefile commands.

This commit aids the development cycle with the exception of linting that will be added in a following release. The README.md and src/Makefile contain the new commands.

mpliego-travelperk committed:
New linting with flake8.

mpliego-travelperk committed:
New Recipe & Ingredient models.

Check src/core/models.py for more information about the created models/tables.

mpliego-travelperk committed:
Fixed missing comment in model tests.

mpliego-travelperk committed:
Registered Recipe & Ingredient to admin.

mpliego-travelperk committed:
Fixed Recipe & Ingredient model relationship.

The intended relationship was by foreingkey rather than ManyToMany join table. This commit fixes that mistake.

mpliego-travelperk committed:
Added Recipe rest api.

This commit includes serializers, views and routes related to the Recipe functionality.

mpliego-travelperk committed:
Fixed typo from double quotes to single quotes.

mpliego-travelperk committed:
Added Ingredient CRUD API.

mpliego-travelperk committed:
Added nested Ingredient in Recipe view.

mpliego committed:
Added CI

Adds pull request branch testing and docker image building for master branch.

mpliego-travelperk committed:
Added documentation regarding the uses of the api.

volker-travelperk commented on src/core/models.py:
```diff
@@ -0,0 +1,34 @@
+        if self.id is None:
+            results = Ingredient.objects.filter(
+                name=self.name, recipe=self.recipe)
+            if len(results) > 0:
```
There are several things here:
1) You could do this query less costy:
```python
if results.exists():
```
this will issue only an `EXISTS` sql query instead of getting the full queryset.

Another option would be to use `results.first()` which will return the first result if there is at least one or return `None`. This could then be used later to assign the id. However,..

2) You have a race condition here if two ingredients are save at the same time. You could add [unique_together](https://docs.djangoproject.com/en/dev/ref/models/options/#django.db.models.Options.unique_together) to the model to make sure at database level that there are no duplicated ingredients per recipe.

3) The model's `save()` method is not always called when you "write" data to the database. For instance `update()` would "circumvent it, so checking this at database level would be better.
However, in the case that you want to have this check in the code, you should add a`select_for_update` to avoid race conditions.

Note: all this goes slightly beyond this exercise, so feel free to not adapt this part of the code. Though, I highly recommend you to read a little about this topic :wink:

mpliego replied:
True, some available options would be creating a lock or based on your recommendation move the responsibility to the DB and use the `unique_together`. 

I moved forward with the `unique_together` and create a migration. Django doesn’t take the `IntegrityError` lightly. I caught the error it in the `add-ingredient` action and disable the creation and update mixings in the ingredient view as there are no immediate use for them. 

Definitely a braking change.

volker-travelperk commented on src/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+Django>=2.1.0,<2.2.0
```
Good practice is to fix the version such that your environment has a determined version no matter when you run/deploy it.

mpliego replied:
Changed!

volker-travelperk commented on src/recipe/views.py:
```diff
@@ -0,0 +1,42 @@
+            return serializers.IngredientSerializer
+        return self.serializer_class
+
+    @action(methods=['POST'], detail=True, url_path='add-ingredient')
```
Did you try to use one of django restframework's mixins?
https://www.django-rest-framework.org/api-guide/generic-views/#createmodelmixin

mpliego replied:
Never thought about it. `@action` was part of the course and use it. I gave it some time boxing probably a cleaner approach in terms of API would be using a nested router https://github.com/alanjds/drf-nested-routers but using an extra dependency was a no go for me.

How would a mixing help in this situation?

volker-travelperk commented on src/recipe/tests/test_recipe.py:
```diff
@@ -0,0 +1,112 @@
+    """Ingredient REST API Tests"""
+
+    def setUp(self):
+        self.client = APIClient()
```
Aren't you missing `super().setUp()`?

mpliego replied:
Good question, Django doesn’t uses it in the [documentation](https://docs.djangoproject.com/en/2.0/topics/testing/advanced/)

If we follow the parent clases the only declaration is in unittest.TestCase
```
def setUp(self):
    "Hook method for setting up the test fixture before exercising it."
    pass
```
Although the setUp method in the test is overriding its parent I don’t see it will be of any use other than following convention.

Let me know your thought.

mpliego-travelperk committed:
Fixed dependency version value.

mpliego-travelperk committed:
Fixed race condition in ingredient save procedure.

In the ingredient model checking for the unique_together condition of the recipe and ingredient name without proper locking created a racing condition. This commit move the responsibility to the database.

mpliego-travelperk committed:
Added CORS endpoints.

mpliego-travelperk committed:
Changed docker package repository url.
